### PR TITLE
feat(backend/store): Add enhanced search functionality for store listings

### PR DESCRIPTION
**Pull request by Swiftyos**
## Summary
- Enhanced the store listing search capability with advanced filtering and improved query building
- Added comprehensive test coverage for the new search functionality
- Optimized database queries for better performance

https://github.com/user-attachments/assets/552ddb83-11f7-4b3c-b6eb-40656fb345e0

## Changes
- **Enhanced search functionality in `db.py`**: 
  - Implemented advanced query building with proper parameter handling
  - Added support for complex filtering options
  - Improved search performance with optimized database queries

- **Added comprehensive test suite**:
  - Created `test_advanced_search.py` with extensive test coverage
  - Tests cover various search scenarios and edge cases
  - Ensures reliability of the enhanced search features

## Test Plan
- [x] All existing tests pass
- [x] New test suite covers all search enhancement scenarios
- [x] Manual testing of search functionality works as expected

🤖 Generated with [Claude Code](https://claude.ai/code)

### CODE REVIEW

File: autogpt_platform/backend/backend/server/v2/store/db.py [hunk 1]
```diff
@@ -23,11 +24,29 @@
...
+
+
+def get_search_service() -> StoreAgentSearchService:
+    """Get or create the search service instance"""
+    global search_service
+    if search_service is None:
+        # Get database URL from environment variable (same as Prisma uses)
+        db_url = os.getenv("DATABASE_URL", "postgresql://localhost:5432")
+        search_service = StoreAgentSearchService(db_url)
+    return search_service
```
Comment (majdyz) [inline autogpt_platform/backend/backend/server/v2/store/db.py:48]: let's use @cache instead of a global variable 

DATABASE_URL should be obtained from db.py; import it directly.

File: autogpt_platform/backend/backend/server/v2/store/db.py [hunk 2]
```diff
@@ -53,6 +72,113 @@ def sanitize_query(query: str | None) -> str | None:
...
+        if query_embedding is None:
+            # Fallback to text-based search if embedding fails
+            logger.warning(
+                f"Failed to create embedding for query: {search_query}. "
+                "Falling back to text search."
+            )
+            return await get_store_agents(
+                search_query=search_query,
+                page=1,
+                page_size=30,
```
Comment (majdyz) [inline autogpt_platform/backend/backend/server/v2/store/db.py:95]: no next page on this?

File: autogpt_platform/backend/backend/server/v2/store/db.py [hunk 3]
```diff
@@ -53,6 +72,113 @@ def sanitize_query(query: str | None) -> str | None:
...
+            page=1,
+            page_size=30,
+        )
+
+    # Convert raw results to StoreAgent models
+    agents = []
+    for row in results:
+        try:
+            # Handle agent_image - it could be a list or a single string
+            agent_image = row.get("agent_image", "")
```
Comment (majdyz) [inline autogpt_platform/backend/backend/server/v2/store/db.py:115]: row can be model_validated into typed model

File: autogpt_platform/backend/backend/server/v2/store/embeddings.py [hunk 4]
```diff
@@ -0,0 +1,469 @@
...
+                )
+                session.add(search_record)
+                search_records.append(search_record)
+
+            await session.commit()
+            return search_records
+
+    async def search_by_embedding(
+        self, query_embedding: List[float], limit: int = 30
+    ) -> List[Dict[str, Any]]:
```
Comment (majdyz) [inline autogpt_platform/backend/backend/server/v2/store/embeddings.py:261]: Do you think this should be a pydantic type return ?

File: autogpt_platform/backend/backend/server/v2/store/embeddings.py [hunk 5]
```diff
@@ -0,0 +1,469 @@
...
+from backend.util.settings import Settings
+
+logger = logging.getLogger(__name__)
+settings = Settings()
+
+# Initialize async OpenAI client with proper configuration
+_openai_client: Optional[AsyncOpenAI] = None
+
+
+def get_openai_client() -> AsyncOpenAI:
```
Comment (majdyz) [inline autogpt_platform/backend/backend/server/v2/store/embeddings.py:26]: same, make it @cached

**Comment by ntindle** [comment]
use something like this in the Prisma schema 

```
search     Unsupported("tsvector")? @default(dbgenerated("''::tsvector"))
```
You'll need to enable partial type generator + a preview feature. Put the tyep generator somewhere useful
```prisma
datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

generator client {
  provider               = "prisma-client-py"
  recursive_type_depth   = 5
  interface              = "asyncio"
  previewFeatures        = ["fullTextSearch"]
  partial_type_generator = "utils/partial_types.py"
}
 ```

An example for Prisma query searching the db + telling it to ensure the return is the correct model type, but this isn't like required or anything just an idea of how to use
```python
async def search_db(
    query: str,
    page: int = 1,
    page_size: int = 10,
    categories: typing.List[str] | None = None,
    description_threshold: int = 60,
    sort_by: str = "rank",
    sort_order: typing.Literal["desc"] | typing.Literal["asc"] = "desc",
    submission_status: prisma.enums.SubmissionStatus = prisma.enums.SubmissionStatus.APPROVED,
) -> typing.List[market.utils.extension_types.AgentsWithRank]:
    """Perform a search for agents based on the provided query string.

    Args:
        query (str): the search string
        page (int, optional): page for searching. Defaults to 1.
        page_size (int, optional): the number of results to return. Defaults to 10.
        categories (List[str] | None, optional): list of category filters. Defaults to None.
        description_threshold (int, optional): number of characters to return. Defaults to 60.
        sort_by (str, optional): sort by option. Defaults to "rank".
        sort_order ("asc" | "desc", optional): the sort order. Defaults to "desc".

    Raises:
        AgentQueryError: Raises an error if the query fails.
        AgentQueryError: Raises if an unexpected error occurs.

    Returns:
        List[AgentsWithRank]: List of agents matching the search criteria.
    """
    try:
        offset = (page - 1) * page_size

        category_filter = ""
        if categories:
            category_conditions = [f"'{cat}' = ANY(categories)" for cat in categories]
            category_filter = "AND (" + " OR ".join(category_conditions) + ")"

        # Construct the ORDER BY clause based on the sort_by parameter
        if sort_by in ["createdAt", "updatedAt"]:
            order_by_clause = f'"{sort_by}" {sort_order.upper()}, rank DESC'
        elif sort_by == "name":
            order_by_clause = f"name {sort_order.upper()}, rank DESC"
        else:
            order_by_clause = 'rank DESC, "createdAt" DESC'

        submission_status_filter = f""""submissionStatus" = '{submission_status}'"""

        sql_query = f"""
        WITH query AS (
            SELECT to_tsquery(string_agg(lexeme || ':*', ' & ' ORDER BY positions)) AS q 
            FROM unnest(to_tsvector('{query}'))
        )
        SELECT 
            id, 
            "createdAt", 
            "updatedAt", 
            version, 
            name, 
            LEFT(description, {description_threshold}) AS description, 
            author, 
            keywords, 
            categories, 
            graph,
            "submissionStatus",
            "submissionDate",
            ts_rank(CAST(search AS tsvector), query.q) AS rank
        FROM "Agents", query
        WHERE 1=1 {category_filter} AND {submission_status_filter}
        ORDER BY {order_by_clause}
        LIMIT {page_size}
        OFFSET {offset};
        """

        results = await prisma.client.get_client().query_raw(
            query=sql_query,
            model=market.utils.extension_types.AgentsWithRank,
        )

        return results

    except prisma.errors.PrismaError as e:
        raise AgentQueryError(f"Database query failed: {str(e)}")
    except Exception as e:
        raise AgentQueryError(f"Unexpected error occurred: {str(e)}")
 ```
You'll need to update the Sql migration to make the search vector
```sql
-- CreateTable
CREATE TABLE "Agents" (
    "id" UUID NOT NULL DEFAULT gen_random_uuid(),
    "createdAt" TIMESTAMP(3) NOT NULL DEFAULT CURRENT_TIMESTAMP,
    "updatedAt" TIMESTAMP(3) NOT NULL,
    "version" INTEGER NOT NULL DEFAULT 1,
    "name" TEXT,
    "description" TEXT,
    "author" TEXT,
    "keywords" TEXT[],
    "categories" TEXT[],
    "search" tsvector DEFAULT ''::tsvector,
    "graph" JSONB NOT NULL,

    CONSTRAINT "Agents_pkey" PRIMARY KEY ("id","version")
);

-- CreateTable
CREATE TABLE "AnalyticsTracker" (
    "id" UUID NOT NULL DEFAULT gen_random_uuid(),
    "agentId" UUID NOT NULL,
    "views" INTEGER NOT NULL,
    "downloads" INTEGER NOT NULL
);

-- CreateIndex
CREATE UNIQUE INDEX "Agents_id_key" ON "Agents"("id");

-- CreateIndex
CREATE UNIQUE INDEX "AnalyticsTracker_id_key" ON "AnalyticsTracker"("id");

-- AddForeignKey
ALTER TABLE "AnalyticsTracker" ADD CONSTRAINT "AnalyticsTracker_agentId_fkey" FOREIGN KEY ("agentId") REFERENCES "Agents"("id") ON DELETE RESTRICT ON UPDATE CASCADE;


-- Add trigger to update the search column with the tsvector of the agent
-- Function to be invoked by trigger

CREATE OR REPLACE FUNCTION update_tsvector_column() RETURNS TRIGGER AS $$

BEGIN

NEW.search := to_tsvector('english', COALESCE(NEW.description, '')|| ' ' ||COALESCE(NEW.name, '')|| ' ' ||COALESCE(NEW.author, ''));

RETURN NEW;

END;

$$ LANGUAGE plpgsql SECURITY definer SET search_path = public, pg_temp;

-- Trigger that keeps the TSVECTOR up to date

DROP TRIGGER IF EXISTS "update_tsvector" ON "Agents";

CREATE TRIGGER "update_tsvector"

BEFORE INSERT OR UPDATE ON "Agents"

FOR EACH ROW

EXECUTE FUNCTION update_tsvector_column ();
 ```
 
extention_types:
```
import prisma.models


class AgentsWithRank(prisma.models.Agents):
    rank: float
 ```
